host-controller/src/main/java/org/jboss/as/host/controller/mgmt/HostControllerRegistrationHandler.java

```diff
@@ -285,9 +307,9 @@ protected HostRegistrationStepHandler(final TransformerRegistry transformerRegis
 
         @Override
         public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
-            // First lock the domain controller
-            context.acquireControllerLock();
-            // Check with the controller lock held
+            // acquires the shared-mode read lock.
+            operationExecutor.beginRegistration(registrationContext.activeOperation.getOperationId(), registrationContext.hostInfo);
+
```

Comment:
After this gets called we have to be sure that completeRegistration gets invoked no matter what happens in the chain of requests.

Actually, why can't it be released in this method via

1) A ResultHandler passed to context.completeStep() and
2) a finally block that calls it if a failure occurs before the context.completeStep() call

That would have the same locking/unlocking semantics as the existing context.acquireControllerLock() (except of course it's the shared lock). The way the OC works with locks is if a step acquires the lock, on the way out after any ResultHandler for that step runs the lock is released by the OC.